### PR TITLE
Add Pihole + Unbound installer

### PR DESCRIPTION
**Pull request by BramSuurdje**
## Description

Pi-hole blocks ads at the DNS level, and Unbound is a secure DNS resolver that validates and caches DNS queries. Combining these tools can significantly improve your online experience and protect your network from potential security threats.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] New Script

### CODE REVIEW

File: install/pihole-unbound-install.sh [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  private-address: 10.0.0.0/8
+  private-address: fd00::/8
+  private-address: fe80::/10
+EOF
+
+cat <<EOF >/etc/dnsmasq.d/99-edns.conf
+edns-packet-max=1232
+EOF
+
+wget https://www.internic.net/domain/named.root -qO- | sudo tee /var/lib/unbound/root.hints
```
Comment (tteck) [inline install/pihole-unbound-install.sh]: `wget -q https://www.internic.net/domain/named.root >/var/lib/unbound/root.hints`

File: install/pihole-unbound-install.sh [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  do-tcp: yes
+  do-ip6: no
+  prefer-ip6: no
+  harden-glue: yes
+  harden-dnssec-stripped: yes
+  use-caps-for-id: no
+  edns-buffer-size: 1232
+  prefetch: yes
+  num-threads: 1
+  private-address: 192.168.0.0/16
```
Comment (tteck) [inline install/pihole-unbound-install.sh]: Wouldn't this be `192.168.0.0/24`?

**Comment by tteck** [comment]
I'm sure that you've tested this successfully?

**Comment by tteck** [comment]
Can Unbound be a selectable option in the Pi-hole LXC that is already in the repository?

**Comment by tteck** [comment]
Something like:

```
read -r -p "Would you like to add Unbound? <y/N> " prompt
if [[ ${prompt,,} =~ ^(y|yes)$ ]]; then
  msg_info "Installing Unbound"
  $STD apt-get install -y unbound
  cat <<EOF >/etc/unbound/unbound.conf.d/pi-hole.conf
server:
  verbosity: 0
  interface: 0.0.0.0
  port: 5335
  do-ip4: yes
  do-udp: yes
  do-tcp: yes
  do-ip6: no
  prefer-ip6: no
  harden-glue: yes
  harden-dnssec-stripped: yes
  use-caps-for-id: no
  edns-buffer-size: 1232
  prefetch: yes
  num-threads: 1
  private-address: 192.168.0.0/24
  private-address: 169.254.0.0/16
  private-address: 172.16.0.0/12
  private-address: 10.0.0.0/8
  private-address: fd00::/8
  private-address: fe80::/10
EOF
  cat <<EOF >/etc/dnsmasq.d/99-edns.conf
edns-packet-max=1232
EOF
  wget -q https://www.internic.net/domain/named.root >/var/lib/unbound/root.hints
  systemctl enable -q --now unbound
  msg_ok "Installed Unbound"
fi
```

**Comment by tteck** [comment]
When you confirm this is functional, I'll merge.

**Comment by BramSuurdje** [comment]
The script now fully works. however i have a detail for you to add to your docs on the website.

- when using Unbound, the user needs to go to the settings page in pihole. and manually set the dns server from google to custom `127.0.0.1#5335`

this may be able to get bypassed if we were using a seperate script for pihole + unbound and setting `PIHOLE_DNS_1` from 8.8.8.8 to `127.0.0.1#5335` in the setupVars.conf, but i haven't tested this.

**Comment by BramSuurdje** [comment]
image

here is a picture for reference

**Comment by BramSuurdje** [comment]
Creating a separate container for Pi-hole and Unbound would increase the container disk space. I remember that Unbound requires a bit more disk space for storing the DNS cache.

**Comment by tteck** [comment]
This solves the issue with changing `PIHOLE_DNS_1` from `8.8.8.8` to `127.0.0.1#5335`

```
read -r -p "Would you like to add Unbound? <y/N> " prompt
if [[ ${prompt,,} =~ ^(y|yes)$ ]]; then
  msg_info "Installing Unbound"
  $STD apt-get install -y unbound
  cat <<EOF >/etc/unbound/unbound.conf.d/pi-hole.conf
server:
  verbosity: 0
  interface: 0.0.0.0
  port: 5335
  do-ip4: yes
  do-udp: yes
  do-tcp: yes
  do-ip6: no
  prefer-ip6: no
  harden-glue: yes
  harden-dnssec-stripped: yes
  use-caps-for-id: no
  edns-buffer-size: 1232
  prefetch: yes
  num-threads: 1
  private-address: 192.168.0.0/16
  private-address: 169.254.0.0/16
  private-address: 172.16.0.0/12
  private-address: 10.0.0.0/8
  private-address: fd00::/8
  private-address: fe80::/10
EOF
  mkdir -p /etc/dnsmasq.d/
  cat <<EOF >/etc/dnsmasq.d/99-edns.conf
edns-packet-max=1232
EOF
  wget -q https://www.internic.net/domain/named.root >/var/lib/unbound/root.hints
  sed -i -e 's/PIHOLE_DNS_1=8.8.8.8/PIHOLE_DNS_1=127.0.0.1#5335/' -e 's/PIHOLE_DNS_2=8.8.4.4/#PIHOLE_DNS_2=8.8.4.4/' /etc/pihole/setupVars.conf
  systemctl enable -q --now unbound
  systemctl restart pihole-FTL.service
  msg_ok "Installed Unbound"
fi
```

**Comment by tteck** [comment]
Can be tested:
```
bash -c "$(wget -qLO - https://github.com/tteck/Proxmox/raw/dev/ct/pihole.sh)"
```

**Comment by tteck** [comment]
With Unbound optimizations:

```
read -r -p "Would you like to add Unbound? <y/N> " prompt
if [[ ${prompt,,} =~ ^(y|yes)$ ]]; then
  msg_info "Installing Unbound"
  $STD apt-get install -y unbound
  cat <<EOF >/etc/unbound/unbound.conf.d/pi-hole.conf
server:
  verbosity: 0
  interface: 0.0.0.0
  port: 5335
  do-ip6: no
  do-ip4: yes
  do-udp: yes
  do-tcp: yes
  num-threads: 1
  hide-identity: yes
  hide-version: yes
  harden-glue: yes
  harden-dnssec-stripped: yes
  harden-referral-path: yes
  use-caps-for-id: no
  harden-algo-downgrade: no
  qname-minimisation: yes
  aggressive-nsec: yes
  rrset-roundrobin: yes
  cache-min-ttl: 300
  cache-max-ttl: 14400
  msg-cache-slabs: 8
  rrset-cache-slabs: 8
  infra-cache-slabs: 8
  key-cache-slabs: 8
  serve-expired: yes
  serve-expired-ttl: 3600
  edns-buffer-size: 1232
  prefetch: yes
  prefetch-key: yes
  target-fetch-policy: "3 2 1 1 1"
  unwanted-reply-threshold: 10000000
  rrset-cache-size: 256m
  msg-cache-size: 128m
  so-rcvbuf: 1m
  private-address: 192.168.0.0/16
  private-address: 169.254.0.0/16
  private-address: 172.16.0.0/12
  private-address: 10.0.0.0/8
  private-address: fd00::/8
  private-address: fe80::/10
EOF
  mkdir -p /etc/dnsmasq.d/
  cat <<EOF >/etc/dnsmasq.d/99-edns.conf
edns-packet-max=1232
EOF
  wget -q https://www.internic.net/domain/named.root >/var/lib/unbound/root.hints
  sed -i -e 's/PIHOLE_DNS_1=8.8.8.8/PIHOLE_DNS_1=127.0.0.1#5335/' -e 's/PIHOLE_DNS_2=8.8.4.4/#PIHOLE_DNS_2=8.8.4.4/' /etc/pihole/setupVars.conf
  systemctl enable -q --now unbound
  systemctl restart pihole-FTL.service
  msg_ok "Installed Unbound"
fi
```

**Comment by tteck** [comment]
https://raw.githubusercontent.com/tteck/Proxmox/dev/install/pihole-install.sh

**Comment by BramSuurdje** [comment]
Tested and working perfectly!

**Comment by tteck** [comment]
Thanks!

**Comment by sunny-azu-red** [comment]
@BramSuurdje, does your configuration use root.hints?

I see your `wget -q https://www.internic.net/domain/named.root >/var/lib/unbound/root.hints` but it does nothing as the file is empty
```console
root@pihole:~# cat /var/lib/unbound/root.hints
root@pihole:~# 
```

[pihole unbound guide](https://docs.pi-hole.net/guides/dns/unbound/) uses this command `wget https://www.internic.net/domain/named.root -qO- | sudo tee /var/lib/unbound/root.hints`

the unbound configuration also doesnt have this line from the guide
```
# Use this only when you downloaded the list of primary root servers!
# If you use the default dns-root-data package, unbound will find it automatically
#root-hints: "/var/lib/unbound/root.hints"
```

thanks!

**Comment by tteck** [comment]
Fixed https://github.com/tteck/Proxmox/commit/3bcf2cece6474889f149cd09048272f7837a35cc

**Comment by sunny-azu-red** [comment]
@tteck not sure if we also need to add the following line into `/etc/unbound/unbound.conf.d/pi-hole.conf` or not?
```
root-hints: "/var/lib/unbound/root.hints"
```

**Comment by tteck** [comment]
https://github.com/tteck/Proxmox/commit/4103d40622b4cf80775381595c9ed4b80f9fdbe3

**Pull request merged**